### PR TITLE
Changes to PagerDuty plugin to support running as a hosted plugin

### PR DESCRIPTION
**Pull request by tcsc**
The major structural change is to provide a new PagerDuty app constructor that allows the auth server to inject a Teleport client, rather than the plugin automaticaly creating a GRPC client on startup.

Other changes include:
 - Added more PagerDuty API types
 - Extra logging
 - Small refactors here & there

### CODE REVIEW

File: api/proto/teleport/legacy/types/types.proto [hunk 1]
```diff
@@ -5172,6 +5172,8 @@ message PluginSpecV1 {
     PluginOktaSettings okta = 4;
     // Settings for device trust jamf plugin
     PluginJamfSettings jamf = 5;
+    // Settings for the PagerDuty plugin
+    PluginPagerDutyAccessSettings pager_duty_access_plugin = 6;
```
Comment (r0mant) [inline api/proto/teleport/legacy/types/types.proto:5176]: Nit: Feels like this is too verbose.
```suggestion
    PluginPagerDutySettings pagerduty = 6;
```

File: api/proto/teleport/legacy/types/types.proto [hunk 2]
```diff
@@ -5197,6 +5199,16 @@ message PluginOpsgenieAccessSettings {
...
 }
 
+message PluginPagerDutyAccessSettings {
+  option (gogoproto.equal) = true;
+
+  // PagerDutyUserEmail is the email address of the PagerDuty user that
+  // will be listed as the reporter source of incidents, comments, etc
+  // within PagerDuty. Should usually be the same user the API key
+  // represents.
+  string pager_duty_user_email = 1;
```
Comment (r0mant) [inline api/proto/teleport/legacy/types/types.proto:5209]:
```suggestion
  string user_email = 1;
```

File: integrations/access/pagerduty/app.go [hunk 3]
```diff
@@ -58,20 +60,46 @@ const (
 // Special kind of error that can be ignored.
 var errSkip = errors.New("")
 
-// App contains global application state.
+// AppSettings hold optional arguments to the PagerDuty plugin app.
+type AppSettings struct {
+	// Teleport is a handle to the client to ue when communicating with
```
Comment (r0mant) [inline integrations/access/pagerduty/app.go:65]:
```suggestion
	// Teleport is a handle to the client to use when communicating with
```

File: integrations/access/pagerduty/app.go [hunk 4]
```diff
@@ -58,20 +60,46 @@ const (
 // Special kind of error that can be ignored.
 var errSkip = errors.New("")
 
-// App contains global application state.
+// AppSettings hold optional arguments to the PagerDuty plugin app.
+type AppSettings struct {
```
Comment (r0mant) [inline integrations/access/pagerduty/app.go:64]: Can't this be made a part of Config? Why a separate struct?

File: integrations/access/pagerduty/client.go [hunk 5]
```diff
@@ -79,27 +80,65 @@ func NewPagerdutyClient(conf PagerdutyConfig, clusterName, webProxyAddr string)
 		}
 	}
 
+	// APIEndpoint parameter is set only in tests
+	endPointURL := "https://api.pagerduty.com"
```
Comment (r0mant) [inline integrations/access/pagerduty/client.go:84]: Nit: Other plugins I think set the default API address as a top-level constant. Maybe move this one out as well for consistency?

File: api/proto/teleport/legacy/types/types.proto (hunk 2 above)
Comment (tcsc) [reply api/proto/teleport/legacy/types/types.proto:5209]: Yeah, I had originally used `user_email` - but I wanted to make it crystal clear that this was the the email for the _PagerDuty_ user, as opposed to the Teleport user,

File: api/proto/teleport/legacy/types/types.proto (hunk 1 above)
Comment (tcsc) [reply api/proto/teleport/legacy/types/types.proto:5176]: Done

File: integrations/access/pagerduty/client.go [hunk 6]
```diff
@@ -79,27 +80,65 @@ func NewPagerdutyClient(conf PagerdutyConfig, clusterName, webProxyAddr string)
...
+	return &types.PluginStatusV1{Code: code}
+}
+
+func onAfterPagerDutyResponse(sink common.StatusSink) resty.ResponseMiddleware {
+	return func(_ *resty.Client, resp *resty.Response) error {
+		log := logger.Get(resp.Request.Context())
+		status := statusFromStatusCode(resp.StatusCode())
+
+		// No usable context in scope, use background with a reasonable timeout
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (marcoandredinis) [inline integrations/access/pagerduty/client.go:135]: Can we extract the timeout to a const?

File: integrations/access/pagerduty/config.go [hunk 7]
```diff
@@ -45,6 +45,7 @@ type PagerdutyConfig struct {
 const NotifyServiceDefaultAnnotation = "pagerduty_notify_service"
 const ServicesDefaultAnnotation = "pagerduty_services"
 
+// TODO(tcsc): re-plant back in `teleport-plugins`
```
[the remark this answers is not in the document]
Comment (tcsc) [reply integrations/access/pagerduty/config.go]: Just deleted

File: integrations/access/pagerduty/client.go [hunk 8]
```diff
@@ -108,17 +147,19 @@ func NewPagerdutyClient(conf PagerdutyConfig, clusterName, webProxyAddr string)
...
 func (p Pagerduty) HealthCheck(ctx context.Context) error {
-	// TODO: decide what to check.
+	var result ListAbilitiesResult
+	_, err := p.client.NewRequest().
+		SetContext(ctx).
+		SetResult(&result).
+		Get("abilities")
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
[the remark this answers is not in the document]
Comment (tcsc) [reply integrations/access/pagerduty/client.go:155]: Yep. Just something that needs authentication to work and is fairly innocuous.

File: integrations/access/pagerduty/client.go [hunk 9]
```diff
@@ -209,6 +249,9 @@ func (p Pagerduty) ResolveIncident(ctx context.Context, incidentID string, resol
 func (p Pagerduty) GetUserInfo(ctx context.Context, userID string) (User, error) {
 	var result UserResult
 
+	p.client.SetDebug(true)
+	defer p.client.SetDebug(false)
+
```
[the remark this answers is not in the document]
Comment (tcsc) [reply integrations/access/pagerduty/client.go:248]: Ooops.

**Review by r0mant — APPROVED** [review]

**Review by marcoandredinis — APPROVED** [review]

**Pull request merged**